### PR TITLE
IP per Container

### PR DESCRIPTION
**Pull request by aquamatthias**
Missing functionality:

- [x] Health Checks need a specific port value instead of portIndex (PR #2731)
- [x] UI Changes
- [x] Add and fill DiscoveryInfo 
  -  [x] Add DiscoveryInfo Protocol Buffers to marathon.proto
  -  [x] Add DiscoveryInfo JSON serialization
  -  [x] Add DiscoveryInfo to AppDefinition.json
  -  [x] Make TaskBuilder fill DiscoveryInfo (PR #2741)
- [ ] MarathonTask can contain an array of container IP's
- [ ] Add Validation rule that makes sure IpAddress and port cannot both be supplied.
- [ ] Testing this feature

### CODE REVIEW

File: src/main/proto/marathon.proto [hunk 1]
```diff
@@ -45,6 +45,12 @@ message HealthCheckDefinition {
   optional bool ignoreHttp1xx = 9 [default = false];
 }
 
+message Network {
+  repeated string groups = 1;
```
Comment (kolloch) [inline src/main/proto/marathon.proto]: The IP address/protocol is missing.

File: src/main/proto/marathon.proto [hunk 2]
```diff
@@ -70,6 +76,7 @@ message ServiceDefinition {
   optional ResourceRoles acceptedResourceRoles = 22;
   optional int64 last_scaling_at = 23;
   optional int64 last_config_change_at = 24;
+  optional Network network = 25;
```
Comment (kolloch) [inline src/main/proto/marathon.proto]: must be repeated

Comment (kolloch) [inline src/main/proto/marathon.proto]: oh, sorry, confused this with `MarathonTask`

File: src/main/scala/mesosphere/marathon/tasks/MarathonTasks.scala [hunk 3]
```diff
@@ -32,4 +32,30 @@ object MarathonTasks {
...
+
+  /**
+    * Returns the host address of the supplied task as a string.
+    *
+    * If the supplied task has at least one NetworkInfo with an IP address
+    * filled in, this function returns the first such address.
+    *
+    * In all other cases, this function returns the slave host address.
+    */
+  def hostAddress(task: MarathonTask): String = {
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/tasks/MarathonTasks.scala]: I am not sure if I like the method name. "host address" has historically meant the IP of the Mesos Agent and not the container network address.

File: src/main/scala/mesosphere/marathon/tasks/MarathonTasks.scala [hunk 4]
```diff
@@ -32,4 +32,30 @@ object MarathonTasks {
...
+        // since Marathon doesn't know how to deal with multiple addresses.
+        task.getNetworksList
+          .asScala
+          .find(_.getIpAddressesCount > 0)
+          .map(_.getIpAddresses(0).getIpAddress)
+      }
+      else None
+
+    // fall back to the slave host address
+    containerAddress getOrElse task.getHost
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/tasks/MarathonTasks.scala]: If you have a new task, you do NOT want to return the agent host address first and then upon the task status update expose the correct IP.

**Comment by gkleiman** [comment]
Superseded by PR #2744 (IP per Task v3)

Leaving open just for reference.

**Comment by kolloch** [comment]
We can reopen it if we need it...

**Pull request closed without merging**